### DISCUSSION TITLE
[Mac Catalyst] Audio recording .net8

### DISCUSSION THREAD

**Discussion by mcocaroUX** [unoplatform/uno#15519] [Q&A] [answered]
After updating to .NET 8, our audio recording functionality broke. While the audio file is created with the correct duration, it produces no sound upon playback.

We're using `AVAudioRecorder` for audio recording. To troubleshoot, I also experimented with `AVCapture` and `AVAudioEngine`, but the outcome remained the same

Here is a snip of our code:

```
AVAudioRecorder _audioRecorder;
NSError _error;

var session = AVAudioSession.SharedInstance();

if(session is not null)
{
    session.SetActive(true);

    session.SetCategory(AVAudioSessionCategory.PlayAndRecord);
}

var settings = new AudioSettings
{
    Format = AudioFormatType.MPEG4AAC,
    SampleRate = 44100.0f,
    NumberChannels = 1,
    AudioQuality = AVAudioQuality.High
};

_audioRecorder = AVAudioRecorder.Create(NSUrl.FromFilename(_filePath), settings, out error);

_audioRecorder.Record();
```

My first assumption was that `AVAudioSession` wasn't selecting the correct device, but is not the case.

Any ideas?

Thanks

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report. All the APIs used in your code sample are coming from .NET, so it would be best to create this issue directly in https://github.com/dotnet/maui/discussions or https://github.com/dotnet/xamarin/xamarin-macios.

**Reply by mcocaroUX** [reply]
Thanks Jerome!

**Reply by sasakrsmanovic** [reply]
@mcocaroUX please let us know the links to those issues when you submit them.

**Comment by ajpinedam** [comment]
Hi @mcocaroUX 

Latest versions of iOS/Mac catalyst require asking for the Mic permission every time you access it. Based on the sample you provided, I think this is what's missing for you.

See the code below. I used this for a sample app and successfully captured audio with .net8.0. The full sample code is here: https://github.com/ajpinedam/UnoAudioGalaxy.


```csharp
//Called from a Buton Click Event
if(AVAudioSession.SharedInstance() is {} session)
{
    session.SetActive(true);
    session.SetCategory(AVAudioSessionCategory.PlayAndRecord);
    session.RequestRecordPermission(HandleMicPermissionCallback);
}
```

```csharp
private void HandleMicPermissionCallback(bool granted)
{
    CoreFoundation.DispatchQueue.MainQueue.DispatchAsync(() =>  //older version do not require using UI thread
    {
        if(!granted)
        {
            Console.WriteLine("Mic Permission denied");
            return;
        }
        StartRecording();
    });
}
```

```csharp
private void StartRecording()
{
    var settings = new AudioSettings
    {
        Format = AudioToolbox.AudioFormatType.MPEG4AAC,
        SampleRate = 12000,
        NumberChannels = 1,
        AudioQuality = AVAudioQuality.High
    };

    _audioRecorder = AVAudioRecorder.Create(NSUrl.FromFilename(_filePath), settings, out var error);

    if(error is not null || _audioRecorder is null)
    {
        Console.WriteLine($"Error creating recorder: {error?.LocalizedDescription}");
        return;
    }

    _audioRecorder.Delegate = new LocalAudioRecorderDelegate();
    _audioRecorder.Record();
}
```

Note: You might also need to update your info.plist file with the new permission, both iOS and Mac Catalyst targets

```xml
<key>NSMicrophoneUsageDescription</key>
<string>Microphone access is required to record an audio.</string>
```

and also the Entitlement.plist 

```xml
<key>com.apple.security.device.audio-input</key>
<true/>
```

**Reply by mcocaroUX** [reply]
@ajpinedam Thank you, I'll try this and let you know
